### PR TITLE
DRAFT - avoid optimization for setuprad.f90 with Intel

### LINKED ISSUE #447: GSI fails to run using newer Intel compilers (i.e, Intel 2022)
The GSI and EnKF compiles without issue using Intel 2022 and the hpc-stack built Intel 2022 libraries.  However, while running the executable built using Intel 2022, the jobs fail:

```
    Start 1: global_T62
1/1 Test #1: global_T62 .................***Failed  240.51 sec
```

Looking at the output:

```
[h1c05:304703:0:304703] Caught signal 11 (Segmentation fault: Sent by the kernel at address (nil))
[h1c05:304708:0:304708] Caught signal 11 (Segmentation fault: Sent by the kernel at address (nil))
```

From the traceback:

```
==== backtrace (tid: 147167) ====
 0 0x000000000004d455 ucs_debug_print_backtrace()  ???:0
 1 0x000000000150cd0d rad_setup_mp_setuprad_()  /scratch1/NCEPDEV/da/Michael.Lueken/test/src/gsi/setuprad.f90:555
 2 0x000000000113c3da gsi_radoper_mp_setup__()  /scratch1/NCEPDEV/da/Michael.Lueken/test/src/gsi/gsi_radOper.F90:100
 3 0x0000000000b0a770 setuprhsall_()  /scratch1/NCEPDEV/da/Michael.Lueken/test/src/gsi/setuprhsall.f90:490
 4 0x000000000111ad0d glbsoi_()  /scratch1/NCEPDEV/da/Michael.Lueken/test/src/gsi/glbsoi.f90:323
 5 0x000000000062e6f7 gsisub_()  /scratch1/NCEPDEV/da/Michael.Lueken/test/src/gsi/gsisub.F90:200
 6 0x000000000040d00d gsimod_mp_gsimain_run_()  /scratch1/NCEPDEV/da/Michael.Lueken/test/src/gsi/gsimod.F90:2209
 7 0x000000000040cf4f MAIN__()  /scratch1/NCEPDEV/da/Michael.Lueken/test/src/gsi/gsimain.f90:631
 8 0x000000000040cee2 main()  ???:0
 9 0x0000000000022555 __libc_start_main()  ???:0
10 0x000000000040cde9 _start()  ???:0
=================================
```

Line 555 in src/gsi/setuprad.f90 is:
https://github.com/NOAA-EMC/GSI/blob/e23204e5ea8f73056ffb2d181ff340deaa0d2163/src/gsi/setuprad.f90#L555

Either compiling the code with -O0 (no optimization) or writing out iuse_rad and predx in setuprad, the same test runs through to completion without a segmentation fault.

To compile using Intel 2022:
 * modulefile/gsi_hera.intel.lua
   * Replace `local hpc_intel_ver=os.getenv("hpc_intel_ver") or "18.0.5.274"`
      * with 
    * `local hpc_intel_ver=os.getenv("hpc_intel_ver") or "2022.1.2"`
   * Replace `local hpc_impi_ver=os.getenv("hpc_impi_ver") or "2018.0.4"`
     * with
   * `local hpc_impi_ver=os.getenv("hpc_impi_ver") or "2022.1.2"`
  * modulefiles/gsi_common.lua
    * Add `local w3nco_ver=os.getenv("w3nco_ver") or "2.4.1"`
    * Add `load(pathJoin("w3nco", w3nco_ver))`

### PR DESCRIPTION
**Pull request by climbfuji**
## Description

This PR is a suggestion/workaround to address the GSI failures with Intel when compiler optimization is enabled. Please ignore the changes to the Orion LUA module, I needed that to (unsuccessfully) attempt running the regression tests myself.

I got far enough to run into a permissions issue for the `global_3dvar` test, because I am not in group `rstprod`. But I couldn't figure out how to run `global_T62`, at least this test doesn't exist when I do
```
cmake -DCONTROLPATH=$PWD/../install/bin -DBUILD_REG_TESTING=ON -DCMAKE_INSTALL_PREFIX=$PWD/../install ../GSI-lower-opt-intel 2>&1 | tee log.cmake
```

This PR may fix https://github.com/NOAA-EMC/GSI/issues/447, but someone needs to test this. @DavidHuber-NOAA?

**Type of change**

Please delete options that are not relevant.

- [x] Bug fix/workaround (non-breaking change which fixes an issue)

**How Has This Been Tested?**

See above
  
**Checklist**

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published

### CODE REVIEW

File: src/gsi/setuprad.f90 [hunk 1]
```diff
@@ -1,4 +1,7 @@
 module rad_setup
+! GSI fails to run using newer Intel compilers (i.e, Intel 2022)
+! with optimization: https://github.com/NOAA-EMC/GSI/issues/447
+!DIR$ OPTIMIZE:0
```
Comment (aerorahul) [inline src/gsi/setuprad.f90]:
```suggestion
! TODO:  Temporary until the root cause is identified and the issue mentioned above is resolved.
!DIR$ OPTIMIZE:0 
```

**Review by aerorahul — COMMENTED** [review]
Thanks @climbfuji 
Just one `TODO` placeholder, so this is not forgotten.

File: src/gsi/setuprad.f90 (hunk 1 above)
Comment (climbfuji) [reply src/gsi/setuprad.f90]: Thanks, @aerorahul . I was also thinking to add something to the top-level README.md / changelog or whatever is appropriate for this repo if this change is acceptable and works as intended.

**Comment by DavidHuber-NOAA** [comment]
Thanks @climbfuji I will give this a try today.

**Comment by DavidHuber-NOAA** [comment]
@climbfuji I gave this a try and the global_3dvar test failed with a crash on gsi/readfiles.f90:724 when compiled with Intel 2022.3.0 on Hera.  Here is the stack trace:

 0 0x000000000004d455 ucs_debug_print_backtrace()  ???:0
 1 0x0000000001e7a09c do_deallocate_all()  for_alloc_copy.c:0
 2 0x00000000012d17c5 read_files_()  /scratch1/NESDIS/nesdis-rdo2/David.Huber/gsi_test/src/gsi/read_files.f90:724
 3 0x0000000000f9d0ae gesinfo_()  /scratch1/NESDIS/nesdis-rdo2/David.Huber/gsi_test/src/gsi/gesinfo.F90:606
 4 0x000000000063a02f gsisub_()  /scratch1/NESDIS/nesdis-rdo2/David.Huber/gsi_test/src/gsi/gsisub.F90:131
 5 0x00000000004164fd gsimod_mp_gsimain_run_()  /scratch1/NESDIS/nesdis-rdo2/David.Huber/gsi_test/src/gsi/gsimod.F90:2266
 6 0x000000000041643f MAIN__()  /scratch1/NESDIS/nesdis-rdo2/David.Huber/gsi_test/src/gsi/gsimain.f90:631
 7 0x00000000004163dd main()  ???:0
 8 0x0000000000022555 __libc_start_main()  ???:0
 9 0x00000000004162f6 _start()  ???:0

I will keep adding the `OPTIMIZE:0` directive to the broken modules until we have something working or it fails to fix an issue.

**Comment by climbfuji** [comment]
Hmm. Something isn't right with this code. Thanks for testing!

**Comment by DavidHuber-NOAA** [comment]
@climbfuji Unfortunately, adding the directive to read_files.F90 did not fix that crash.  On inspection, I noticed that this crash occurs on the last line of the module, according to the backtrace, so I went up another level and tried deoptimizing gesinfo.F90, which also didn't work; same crash.  This work is in [port_2_i2021](https://github.com/DavidHuber-NOAA/GSI/tree/port_2_i2021).

Reading up on the directive itself, setting optimization flags is a 'suggestion' to the compiler and may be ignored without any diagnostic output.  So instead, I'm going to modify the CMake file to set `-O0` for these three modules.

**Comment by climbfuji** [comment]
> @climbfuji Unfortunately, adding the directive to read_files.F90 did not fix that crash. On inspection, I noticed that this crash occurs on the last line of the module, according to the backtrace, so I went up another level and tried deoptimizing gesinfo.F90, which also didn't work; same crash. This work is in [port_2_i2021](https://github.com/DavidHuber-NOAA/GSI/tree/port_2_i2021).
> 
> Reading up on the directive itself, setting optimization flags is a 'suggestion' to the compiler and may be ignored without any diagnostic output. So instead, I'm going to modify the CMake file to set `-O0` for these three modules.

Thanks for all the tests, @DavidHuber-NOAA. This is very annoying, but if instead of using directives setting cmake compiler flags works, then that's fine!

**Comment by DavidHuber-NOAA** [comment]
@climbfuji Unfortunately, deoptimizing the 3 modules still resulted in a segmentation fault in read_files.f90.  I took it another step and deoptimized every module that was called by read_files.f90 and gesinfo.F90, but I am still getting the same seg fault.

**Comment by climbfuji** [comment]
Closing this, since we have a better solution in #447

**Pull request closed without merging**